File: Shared09/Code/Calculator.c
#include <stdio.h>

// Single Responsibilty Design
int sum(int x, int y) { return x + y; }
int sub(int x, int y) { return x - y; }
int mul(int x, int y) { return x * y; }
int div(int x, int y) { return x / y; }

// Polymorphism
// Mechanim 03: Passing Behaviour To A Behaviour

// Single Responsibilty Design
// Open For Extension And Close Modification
int calculator(int x, int y, int (*operation)(int, int) ) {
	return operation(x, y);
}

int main() {
	int a = 10, b = 20;
	int result = 0;

	result = calculator(a, b, sum);
	printf("\nResult : %d", result);

	result = calculator(a, b, sub);
	printf("\nResult : %d", result);

	result = calculator(a, b, mul);
	printf("\nResult : %d", result);

	result = calculator(a, b, mul);
	printf("\nResult : %d", result);
}

File: Shared09/Code/Calculator.kt

package learnKotlin;

fun sum(x: Int, y: Int) : Int { return x + y }
fun sub(x: Int, y: Int) : Int { return x - y }
fun sum3(x: Int, y: Int, z: Int) : Int { return x + y + z }

fun calculator(x: Int, y: Int, operation: (Int, Int) -> Int ) : Int { 
	return operation(x, y)
}

fun main() {
	val a = 10
	val b = 20
	var result: Int

	result = calculator(a, b, ::sum)
	println("Result : $result")

	result = calculator(a, b, ::sub)
	println("Result : $result")

	// What Is The Type of something variable
	var something: (Int, Int) -> Int = ::sum
	something = ::sum3 // (Int, Int, Int) -> Int

	result = something(100, 200)
	println("Result : $result")

}
